### PR TITLE
Standardize on using `lubridate::leap_year` 

### PR DESCRIPTION
**Pull request by bpbond**
Wow…everyone had rolled their own (generally incorrect) way of computing leap years. Tried to search and catch everything.

### CODE REVIEW

**Comment by bpbond** [comment]
I also like the way folks had defined a `paste0` function in R. Good idea!

File: models/ed/inst/timeutils.R [hunk 1]
```diff
@@ -17,36 +17,13 @@
...
-
-
-
-
-#==========================================================================================#
-#==========================================================================================#
 #      Function that determines the number of days in a given month.                       #
 #------------------------------------------------------------------------------------------#
 daymax = function(month,year){
   mmm  = c(31,28,31,30,31,30,31,31,30,31,30,31)
```
Comment (robkooper) [inline models/ed/inst/timeutils.R]: or use

```
daymax = function(month,year) {
  return(as.integer(days_in_month(as.Date(paste(year, month, 1, sep='-')))))
}
```

File: modules/assim.batch/R/get.da.data.R [hunk 2]
```diff
@@ -115,7 +115,7 @@ get.da.data <- function(out.dir, ameriflux.dir, years, be, bu, ensemble.size=199
   points.per.day <- 48
   dtime <- do.call(c, lapply(years,
           function(year) year + 
-                seq(1, (366 + (year%%4==0)), by=1/points.per.day)[-1] / 366))
+                seq(1, (366 + lubridate::leap_year(year)), by=1/points.per.day)[-1] / 366))
```
Comment (robkooper) [inline modules/assim.batch/R/get.da.data.R]: should this not be 365+lubridate::leap_year(year)

File: modules/data.atmosphere/inst/scripts/flux2lsm.v2.R [hunk 3]
```diff
@@ -115,15 +115,13 @@ AirDens <- function(pres, T, rv){
 firstday <- function(mo,yr){
   ldoy = c(0,31,60,91,121,152,182,213,244,274,305,335,366)  
   doy = c(0,31,59,90,120,151,181,212,243,273,304,334,365)  
-  if(yr%%4 == 0){return(ldoy[mo])}
-  return(doy[mo])
+  ifelse(lubridate::leap_year(yr), ldoy[mo], doy[mo])
 }
 
 nday <- function(mo,yr){
```
Comment (robkooper) [inline modules/data.atmosphere/inst/scripts/flux2lsm.v2.R]: same as daymax function

File: modules/data.atmosphere/inst/scripts/NARRinventory.R [hunk 4]
```diff
@@ -5,11 +5,9 @@ num <- function(x){as.numeric(as.character(x))}
 mchar <- c("01","02","03","04","05","06","07","08","09","10","11","12")
 mlennorm <- c(31,28,31,30,31,30,31,31,30,31,30,31)
 mlenleap <- c(31,29,31,30,31,30,31,31,30,31,30,31)
-mlen <- function(m,y){
-  if(y%%4 == 0){ ##leap year
-    return(mlenleap[m])
-  }
-  return(mlennorm[m])  
+
+mlen <- function(m, y) {
```
Comment (robkooper) [inline modules/data.atmosphere/inst/scripts/NARRinventory.R]: same as daymax function

File: modules/data.atmosphere/inst/scripts/flux2lsm.v2.R [hunk 5]
```diff
@@ -115,15 +115,13 @@ AirDens <- function(pres, T, rv){
 firstday <- function(mo,yr){
   ldoy = c(0,31,60,91,121,152,182,213,244,274,305,335,366)
```
Comment (robkooper) [inline modules/data.atmosphere/inst/scripts/flux2lsm.v2.R]: can use:

```
yday(as.Date(paste(yr, mo, 1, sep="1")))
```

File: modules/assim.batch/R/get.da.data.R (hunk 2 above)
Comment (bpbond) [inline modules/assim.batch/R/get.da.data.R]: Good catch.

**Comment by bpbond** [comment]
Thanks @robkooper for the good suggestions and bug catch.

**Comment by mdietze** [comment]
@robkooper since you've looked at some of this already, could you do the code review?

**Comment by serbinsh** [comment]
Ben,

When you say defined paste0, what do you mean? That is a built in function isn't it?

S

________________________________

**Comment by robkooper** [comment]
@serbinsh see https://github.com/PecanProject/pecan/pull/1064/files#diff-fa2505c121632f413c683d9b71a1d38aL39

**Comment by bpbond** [comment]
@serbinsh - what @robkooper said. Not trying to slag on anyone, but it made me laugh.

**Comment by serbinsh** [comment]
Ha!  OK, I see now.  Sort of what I was going for, that is not re-creating a built in

________________________________

File: modules/assim.batch/R/get.da.data.R (hunk 2 above)
Comment (robkooper) [inline modules/assim.batch/R/get.da.data.R]: @bpbond can you change this?

**Comment by bpbond** [comment]
@robkooper in your request above, should the divisor (`/366`) stay at 366? Or go to 365? Want to double-check the logic of what's happening... Thx.

**Comment by robkooper** [comment]
My gut feeling is:

```
nodays <- 365 + lubridate::leap_year(year)
seq(1, nodays, by=1/points.per.day)[-1] / nodays))
```

**Comment by bpbond** [comment]
That was my first thought too.

File: modules/assim.batch/R/get.da.data.R [hunk 6]
```diff
@@ -114,8 +114,10 @@ get.da.data <- function(out.dir, ameriflux.dir, years, be, bu, ensemble.size=199
 
   points.per.day <- 48
   dtime <- do.call(c, lapply(years,
-          function(year) year + 
-                seq(1, (366 + (year%%4==0)), by=1/points.per.day)[-1] / 366))
+          function(year) {
+            nodays <- 365 + lubridate::leap_year(year)
+            year + seq(1, nodays, by = 1 / points.per.day)[-1] / nodays))
```
Comment (tonygardella) [inline modules/assim.batch/R/get.da.data.R]: extra parenthesis here is throwing travis off. ```year + seq(1, nodays, by = 1 / points.per.day)[-1] / nodays))``` should be ```year + seq(1, nodays, by = 1 / points.per.day)[-1] / nodays)```

File: modules/data.atmosphere/R/download.NLDAS.R [hunk 7]
```diff
@@ -51,25 +51,25 @@ download.NLDAS <- function(outfolder, start_date, end_date, site_id, lat.in, lon
...
       day1 <- yday(start_date)
       days.use = day1:nday
       nday=length(days.use) # Update nday
     } else if(i==rows) {
       # If this is the last of many years, we only need to worry about the start date
-      nday  = ifelse(lubridate:: leap_year(year), 366, 365) # leap year or not; days per year
+      nday  = ifelse(lubridate::leap_year(year), 366, 365) # leap year or not; days per year
       day2 <- yday(end_date)
       days.use = 1:day2
       nday=length(days.use) # Update nday
```
Comment (ashiklom) [inline modules/data.atmosphere/R/download.NLDAS.R]: This is just screaming to be broken. Can we pull the common definition (`nday`) out of the `if` statements? Or better yet, make this whole thing a single function that gets called 4 times?

File: modules/data.atmosphere/R/metgapfill.R [hunk 8]
```diff
@@ -146,13 +146,13 @@ metgapfill <- function(in.path, in.prefix, outfolder, start_date, end_date, lst=
...
            dt <- (365*24*60*60)/length(sec)) #non-leap year
-    ifelse(leap_year(year)==TRUE,
+    ifelse(leap_year(year),
            doy <- rep(1:366,each=86400/dt),
            doy <- rep(1:365,each=86400/dt))
-    ifelse(leap_year(year)==TRUE,
+    ifelse(leap_year(year),
            hr <- rep(seq(0,length=86400/dt,by=dt/86400*24),366),
            hr <- rep(seq(0,length=86400/dt,by=dt/86400*24),365))
     f <- pi/180*(279.5+0.9856*doy)
```
Comment (ashiklom) [inline modules/data.atmosphere/R/metgapfill.R]: This seems like a confusing and roundabout way to do this. Wouldn't it better to have:
```r
if (leap_year(year)) {
  dt <- ...
  doy <- ...
  hr <- ...
} else {
  dt <- ...
  doy <- ...
  hr <- ...
}

File: modules/data.atmosphere/R/download.NLDAS.R (hunk 7 above)
Comment (ashiklom) [inline modules/data.atmosphere/R/download.NLDAS.R]: Also, should we just define some functions in `utils` or something for getting the number of days, hours, seconds, etc. in a year? E.g.

```r
nday <- function(year) {
  ifelse(lubridate::leap_year(year), 366, 365)
}

nhour <- function(year) {
  ifelse(lubridate::leap_year(year), ...)
}
```

File: modules/data.atmosphere/R/metgapfill.R (hunk 8 above)
Comment (ashiklom) [inline modules/data.atmosphere/R/metgapfill.R]: Also, see my comment above about `nday`, `nhour` functions.

File: web/sugarcane/sugarcaneForBioCro.R [hunk 9]
```diff
@@ -164,16 +164,13 @@ library(RNCEP)
...
        indx<-as.integer(length(day))
        day[as.integer(indx+1):as.integer(indx+366)]=seq(1:366)
        year[as.integer(indx+1):as.integer(indx+366)]=rep(i,366)
-     }
-      if(!((i%%400)==0 || (i%%100!=0 && i%%4==0)))
-        {
+     } else {
        indx<-as.integer(length(day))
        day[as.integer(indx+1):as.integer(indx+365)]=seq(1:365)
        year[as.integer(indx+1):as.integer(indx+365)]=rep(i,365)
```
Comment (ashiklom) [inline web/sugarcane/sugarcaneForBioCro.R]: Per my comment above about `nday` as a function, we could then combine code like this into just:

```r
day[as.integer(indx+1):as.integer(indx + nday(i))] = seq(1:nday(1))
```

File: modules/assim.batch/R/get.da.data.R [hunk 10]
```diff
@@ -114,8 +114,10 @@ get.da.data <- function(out.dir, ameriflux.dir, years, be, bu, ensemble.size=199
 
   points.per.day <- 48
   dtime <- do.call(c, lapply(years,
-          function(year) year + 
-                seq(1, (366 + (year%%4==0)), by=1/points.per.day)[-1] / 366))
+          function(year) {
+            nodays <- 365 + lubridate::leap_year(year)
+            year + seq(1, nodays, by = 1 / points.per.day)[-1] / nodays
+          }))
```
Comment (ashiklom) [inline modules/assim.batch/R/get.da.data.R]: Why not just use `sapply`?

File: models/ed/inst/timeutils.R [hunk 11]
```diff
@@ -17,48 +17,16 @@
...
-  mday = mmm[month]
-
-  addone       = month == 2 & is.leap(year)
-  mday[addone] = mday[addone] + 1
-
-  return(mday)
-} #end function
+daymax = function(month, year) {
+  as.integer(days_in_month(as.Date(paste(year, month, 1, sep='-'))))
+} # daymax
```
Comment (ashiklom) [inline models/ed/inst/timeutils.R]: What about `lubridate::days_in_month`?

Comment (bpbond) [inline models/ed/inst/timeutils.R]: Yeah, I had marked this as something to come back to, but didn't want to unnecessarily mess with others' code.

**Comment by bpbond** [comment]
I don't think this is holding on me, is it? @robkooper you requested changes but I think everything in my bailiwick is done?

**Comment by robkooper** [comment]
@mdietze biggest questions still stands

Is it correct to use the following
```
nodays <- 365 + lubridate::leap_year(year)
seq(1, nodays, by=1/points.per.day)[-1] / nodays))
```

**Comment by mdietze** [comment]
@robkooper that looks correct to me. Thought I replied on Gitter last week, sorry if that got overlooked.

@bpbond there's a file conflict on get.da.data.R

**Pull request merged**